File: heap/leetcode_2231/solution.py
class Solution:
    def largestInteger(self, num: int) -> int:
        digits = [int(d) for d in str(num)]

        odd_digits = sorted([d for d in digits if d % 2 == 1], reverse=True)
        even_digits = sorted([d for d in digits if d % 2 == 0], reverse=True)
        result = []

        odd_index = 0
        even_index = 0

        for d in digits:
            if d % 2 == 0:
                result.append(even_digits[even_index])
                even_index += 1
            else:
                result.append(odd_digits[odd_index])
                odd_index += 1

        return int("".join(map(str, result)))

File: heap/leetcode_2231/solution.js
/**
 * @param {number} num
 * @return {number}
 */
var largestInteger = function(num) {
    const digits = Array.from(String(num), Number);

    const oddDigits = [];
    const evenDigits = [];
    const result = [];
    
    for(const digit of digits) {
        if (digit % 2 === 0){
            evenDigits.push(digit)
        } else {
            oddDigits.push(digit)
        }
    }

    oddDigits.sort((a, b) => b - a);
    evenDigits.sort((a, b) => b - a);

    let oddIndex = 0;
    let evenIndex = 0;


    for (let i = 0; i < digits.length; i++) {
        if (digits[i] % 2 === 0) {
            result[i] = evenDigits[evenIndex++];
        } else {
            result[i] = oddDigits[oddIndex++];
        }
    }

    return parseInt(result.join(''));
};